### PR TITLE
BSR optimization in TOPI

### PR DESCRIPTION
**Pull request by elphinkuo**
Add BSRMM and BSRMV in TOPI of Python layer.
Add BSRNDArray in contrib of Python Python layer.

This is the first submission of a series work, to support the BSR optimization in TVM. The first submission is in draft PR mode, as we are trying to follow the design principles and code styles of TVM. If there is any issue or modification we need to address, feel free to let us know.
More details are in this paper https://arxiv.org/abs/2106.08846.

@tqchen @yzhliu @vinx13

### CODE REVIEW

File: python/tvm/topi/bsrmm.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    ), "Data and weight must have the same dtype, but they have %s and %s" % (
+        data.dtype,
+        weight.dtype,
+    )
+    if bias is not None:
+        assert len(bias.shape) == 1
+    M = simplify(indptr.shape[0] - 1)
+    _, N = weight.shape
+
+    def bsrmm_default_ir(data, indices, indptr, blocksize, weight, out):
```
Comment (vinx13) [inline python/tvm/topi/bsrmm.py:77]: if this implementation is target specific, would be better to put it under target's directory e.g https://github.com/apache/tvm/blob/main/python/tvm/topi/x86/sparse.py

**Comment by tkonolige** [comment]
Hi @elphinkuo, I implemented sparse matrix multiplication for the GPU in TVM. I can help review and provide feedback. You've marked this PR as a draft, do you want it reviewed right now or should I wait?

File: python/tvm/topi/bsrmm.py (hunk 1 above)
Comment (elphinkuo) [reply python/tvm/topi/bsrmm.py:77]: Got it! We will take your point into consideration. 
Thanks a lot for the prompt response!

**Comment by elphinkuo** [comment]
> Hi @elphinkuo, I implemented sparse matrix multiplication for the GPU in TVM. I can help review and provide feedback. You've marked this PR as a draft, do you want it reviewed right now or should I wait?

Thanks a lot for the quick reply! @tkonolige   We may improve this PR shortly and it is good to review at that time.

**Comment by jroesch** [comment]
@elphinkuo can you update this PR?

**Pull request closed without merging**